ext/curl/share.c

```diff
@@ -134,6 +136,112 @@ PHP_FUNCTION(curl_share_strerror)
 }
 /* }}} */
 
+/**
+ * Initialize a persistent curl share handle with the given share options, reusing an existing one if found.
+ *
+ * Throws an exception if the share options are invalid.
+ */
+PHP_FUNCTION(curl_persistent_share_init)
+{
+	zval *share_opts = NULL, *entry = NULL;
+	zend_ulong persistent_id = 0;
+
+	php_curlsh *sh;
+
+	CURLSHcode error;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY_EX(share_opts, 0, 1)
+	ZEND_PARSE_PARAMETERS_END();
+
+	object_init_ex(return_value, curl_persistent_share_ce);
+	sh = Z_CURL_SHARE_P(return_value);
+
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(share_opts), entry) {
+		ZVAL_DEREF(entry);
+
+		zend_ulong option = zval_get_long_ex(entry, true);
+
+		if (option == CURL_LOCK_DATA_COOKIE) {
+			zend_throw_exception_ex(
+				NULL,
+				0,
+				"CURL_LOCK_DATA_COOKIE is not allowed with persistent curl share handles"
+			);
+
+			goto error;
+		}
+
+		// Ensure that each additional option results in a unique persistent ID.
+		persistent_id += 1 << option;
+	} ZEND_HASH_FOREACH_END();
+
+	zend_array_sort(Z_ARRVAL_P(share_opts), php_array_data_compare_unstable_i, 1);
```

Comment:
Not sure what use the sorting is here, because you already calculated the `persistent_id`.

Reply:
I'm sorting this so that I can set it as the `$options` class property on the line below (L180); I did not want the order of the property to be dependent on the order of the argument array.

Reply:
Personally, I don't see much issue with this dependency. Do you have any specific example in mind where this would cause issues? And I guess by removing this line, there would be no need to separate the input array (?).